server/src/main/java/io/crate/planner/operators/Get.java

```diff
@@ -135,20 +147,55 @@ public ExecutionPlan build(PlannerContext plannerContext,
                 .orElse(SequenceNumbers.UNASSIGNED_PRIMARY_TERM);
             pkAndVersions.add(new PKAndVersion(id, version, sequenceNumber, primaryTerm));
         }
-        return new Collect(
+
+        var docKeyColumns = new ArrayList<>(docTableInfo.primaryKey());
+        docKeyColumns.addAll(docTableInfo.partitionedBy());
+        docKeyColumns.add(docTableInfo.clusteredBy());
+        docKeyColumns.add(DocSysColumns.VERSION);
+        docKeyColumns.add(DocSysColumns.SEQ_NO);
+        docKeyColumns.add(DocSysColumns.PRIMARY_TERM);
+
+        List<Symbol> boundOutputs = Lists2.map(
+            outputs, s -> SubQueryAndParamBinder.convert(s, params, subQueryResults));
+        var boundQuery = SubQueryAndParamBinder.convert(query, params, subQueryResults);
+
+        var toCollect = new LinkedHashSet<>(boundOutputs);
+        Consumer<Reference> addRefIfMatch = ref -> {
+            if (ref.ident().tableIdent().equals(tableRelation.relationName())
+                && docKeyColumns.contains(ref.column()) == false) {
```

Comment:
I also don't quite understand why this check is necessary. `toCollect` is a set, so duplicates won't be added anyways?

Is this under the assumption that the symbol might not be needed because it is only used in the docKeys part? I'm not sure if that assertion is safe. Couldn't the query still use columns from `docKeyColumns` ?

Reply:
Maybe it would make sense to use something like `SymbolVisitors.intersection(query, allPossibleTableOutputs, toCollect::add);` ?

Reply:
Yep right, good catch. This logic which is intended to only add a filter projection when needed (other columns than DocKeys are part of the query) was not correct. I'll push a new safe logic.

Reply:
Push a fixup before I saw your recent suggestion with the intersection. If you still think I should go in this direction instead, please shout.

Reply:
And related to `WHERE .. (pk = 1 AND substr(x, ..)`, as the `RevVisitor` traverses the tree `x` will be added to the outputs. This query works as expected. Do I miss anything?

Reply:
I thought something along the lines of

```
create table tbl (id int primary key, x int);
select x from tbl where id = 1 and (_seq_no = 1 or x = 1)
```

But that is forbidden.

Or 

```
select x from tbl where id = 1 and (id = 3 or x = 1);
```

but that works. So all good :+1: